src/chainstate/stacks/boot/cost-voting.clar

```diff
@@ -0,0 +1,263 @@
+    (map-get? proposals { proposal-id: proposal-id }))
+
+;; getter for confirmed cost-function proposals
+(define-read-only (get-confirmed-proposal (confirmed-id uint))
+    (map-get? confirmed-proposals { confirmed-id: confirmed-id }))
+
+;; getter for cost-function proposal votes
+(define-read-only (get-proposal-votes (proposal-id uint))
+    (get votes (map-get? proposal-votes { proposal-id: proposal-id })))
+
+;; getter for cost-function proposal vetos
+(define-read-only (get-proposal-vetos (proposal-id uint))
+    (get vetos (map-get? proposal-vetos { proposal-id: proposal-id })))
+
+;; getter for cost-function proposal votes, for specific principal
+(define-read-only (get-principal-votes (address principal) (proposal-id uint))
+    (get votes (map-get? principal-proposal-votes { address: address, proposal-id: proposal-id })))
+
+;; Propose cost-functions
+(define-public (submit-proposal (function-contract principal)
+                                (function-name (string-ascii 128))
+                                (cost-function-contract principal)
+                                (cost-function-name (string-ascii 128)))
+    (begin
+        (map-insert proposals { proposal-id: (var-get proposal-count) }
+                              { cost-function-contract: cost-function-contract,
+                                cost-function-name: cost-function-name,
+                                function-contract: function-contract,
+                                function-name: function-name,
+                                expiration-block-height: (+ burn-block-height VOTE_LENGTH) })
+        (map-insert proposal-votes { proposal-id: (var-get proposal-count) } { votes: u0 })
+        (var-set proposal-count (+ (var-get proposal-count) u1))
+        (ok (- (var-get proposal-count) u1))))
+
+;; Vote on a proposal
+(define-public (vote-proposal (proposal-id uint) (amount uint))
+    (let (
+        (expiration-block-height (get expiration-block-height (unwrap! (map-get? proposals {
+            proposal-id: proposal-id }) (err ERR_NO_SUCH_PROPOSAL))))
+        (cur-votes (default-to u0 (get votes (map-get? proposal-votes { proposal-id: proposal-id }))))
+        (cur-principal-votes (default-to u0 (get votes (map-get? principal-proposal-votes {
+            address: tx-sender,
+            proposal-id: proposal-id }))))
+    )
+
+    ;; a vote must have a positive amount
+    (asserts! (> amount u0) (err ERR_AMOUNT_NOT_POSITIVE))
+
+    ;; the vote must occur before the expiration
+    (asserts! (< burn-block-height expiration-block-height) (err ERR_PROPOSAL_EXPIRED))
+
+    ;; the proposal must not already be voter confirmed
+    (asserts! (is-none (map-get? vote-confirmed-proposals { proposal-id: proposal-id }))
+        (err ERR_VOTE_ENDED))
+
+    (unwrap! (stx-transfer? amount tx-sender (as-contract tx-sender)) (err ERR_INSUFFICIENT_FUNDS))
+    (unwrap! (ft-mint? cost-vote-token amount tx-sender) (err ERR_UNREACHABLE))
```

Comment:
Can you briefly explain why there's a fungible voting token?  I can understand the need to ensure that STX deposits and withdrawals are necessary, but why is a data map insufficient for that?

Also, `unwrap-panic` can be used in place of having a `(err ERR_UNREACHABLE)` if you want to have the whole transaction abort.  A good reason to use an `ERR_UNREACHABLE`-like error is if you want to trigger some kind of error (or panic) in the VM itself, whereby the VM intercepts the result and checks the error message.  Is that going to happen here?

Reply:
The FT allows users to apply post-conditions to prevent inadvertent withdrawals.

Reply:
Also, having a dedicated voting token allows you to prevent surreptitious withdrawals by using the "deny" mode in post-conditions -- unless you intended the voting token to move, any attempt to withdraw your STX from the vote will cause the transaction to abort.